### PR TITLE
refactor(build): convert to typescript

### PR DESCRIPTION
**Pull request by caugner**
## Summary

Converts `build/` to TypeScript.

### CODE REVIEW

**Comment by caugner** [comment]
FYI There are (hopefully) just some type errors in `build/cli.ts` left:

https://github.com/mdn/yari/runs/7238162391?check_suite_focus=true#step:9:18

File: build/sync-translated-content.ts [hunk 1]
```diff
@@ -237,20 +237,14 @@ function syncTranslatedContent(inFilePath, locale) {
...
 
-function syncTranslatedContentForAllLocales() {
+export function syncTranslatedContentForAllLocales() {
   let moved = 0;
   for (const locale of VALID_LOCALES.keys()) {
     if (locale == "en-us") {
       continue;
     }
-    const { stats: { movedDocs = 0 } = {} } = syncAllTranslatedContent(locale);
+    const { movedDocs = 0 } = syncAllTranslatedContent(locale);
```
Comment (caugner) [inline build/sync-translated-content.ts:246]: Looks like a bug.

Comment (caugner) [reply build/sync-translated-content.ts:246]: Hm, it's actually a bug fix, but let's back it out.

File: client/src/page-not-found/fallback-link.tsx [hunk 2]
```diff
@@ -2,10 +2,14 @@ import React from "react";
...
 
-import { Doc } from "../../../libs/types/document";
-import LANGUAGES_RAW from "../../../libs/languages";
+import { Doc, LanguageItem } from "../../../libs/types";
 import NoteCard from "../ui/molecules/notecards";
 
+const LANGUAGES_RAW = require("../../../libs/languages") as Record<
+  string,
+  LanguageItem
+>;
```
Comment (LeoMcA) [inline client/src/page-not-found/fallback-link.tsx:11]: Could we move this (and the same in other files) into something like `libs/languages/index.ts` so we can `import` in typescript, while continuing to `require()` in js?

**Review by LeoMcA — APPROVED** [review]
looks good!

just some small observations, nothing major - feel free to merge once you've addressed them (or confirmed you didn't mean to do the TODOs in this PR)

File: build/index.ts [hunk 3]
```diff
@@ -1,36 +1,48 @@
...
+export { buildSPAs } from "./spas";
+import { renderCache as renderKumascriptCache } from "../kumascript";
+const LANGUAGES_RAW = require("../libs/languages") as Record<
+  string,
+  LanguageItem
+>;
+import { safeDecodeURIComponent } from "../kumascript/src/api/util";
+import { wrapTables } from "./wrap-tables";
+
+// TODO Deduplicate.
```
[the remark this answers is not in the document]
Comment (caugner) [reply build/index.ts]: Fixed in bf9b4741e.

File: build/index.ts [hunk 4]
```diff
@@ -1,36 +1,48 @@
...
-const { FLAW_LEVELS } = require("../libs/constants");
-const {
+import { Doc, Flaws, LanguageItem } from "../libs/types";
+import fs from "fs";
+import path from "path";
+
+import chalk from "chalk";
+import * as cheerio from "cheerio";
+import {
+  MacroDeprecatedError,
```
[the remark this answers is not in the document]
Comment (caugner) [reply build/index.ts]: Fixed in dacff3d54.

File: build/index.ts [hunk 5]
```diff
@@ -1,36 +1,48 @@
...
+  MacroLiveSampleError,
+  MacroRedirectedLinkError,
+  SourceCodeError,
```
[the remark this answers is not in the document]
Comment (caugner) [reply build/index.ts]: Fixed in dacff3d54.

File: build/index.ts [hunk 6]
```diff
@@ -611,10 +661,17 @@ async function buildDocument(document, documentOptions = {}) {
     document.metadata.slug.startsWith("orphaned/") ||
     document.metadata.slug.startsWith("conflicting/");
 
-  return { doc, liveSamples, fileAttachments, bcdData };
+  // TODO Refactor to avoid `as Doc` type assertion.
```
[the remark this answers is not in the document]
Comment (caugner) [reply build/index.ts:649]: Removed in 51f156462.

File: build/spas.ts [hunk 7]
```diff
@@ -252,13 +268,11 @@ async function buildSPAs(options) {
...
 
-      let featuredContributor = null;
-      if (contributorSpotlightRoot) {
-        featuredContributor = await buildContributorSpotlight(locale, options);
-      }
+      let featuredContributor = contributorSpotlightRoot
+        ? await buildContributorSpotlight(locale, options)
+        : null;
 
       // circular dependency, so needs to be imported down here:
```
[the remark this answers is not in the document]
Comment (caugner) [reply build/spas.ts]: Removed in 26597590b.

File: client/src/page-not-found/fallback-link.tsx (hunk 2 above)
Comment (caugner) [reply client/src/page-not-found/fallback-link.tsx]: Fixed in 7e75de5e4.

PS: This is possible because of: https://github.com/mdn/yari/pull/7006

**Pull request merged**